thread-rcu/rcu.h

```diff
@@ -0,0 +1,246 @@
+/* Target to use Linux Kernel Memory Model (LKMM) for thread-rcu,
+ * C11 memory model may no compatible to LKMM.
+ * Be careful about the architecture you use.
+ * you can check the paper to see more detail:
+ *
+ * http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0124r6.html
+ */
+
+#ifndef __RCU_H__
+#define __RCU_H__
+
+/* lock primitives from pthread and compiler primitives */
+
+#include <pthread.h>
+#include <stdatomic.h>
+
+typedef pthread_mutex_t spinlock_t;
+
+#define SPINLOCK_INIT PTHREAD_MUTEX_INITIALIZER
+static inline void spin_lock(spinlock_t *sp)
+{
+    int ret;
+
+    ret = pthread_mutex_lock(sp);
+    if (ret != 0) {
+        fprintf(stderr, "spin_lock:pthread_mutex_lock %d\n", ret);
+        abort();
+    }
+}
+
+static inline void spin_unlock(spinlock_t *sp)
+{
+    int ret;
+
+    ret = pthread_mutex_unlock(sp);
+    if (ret != 0) {
+        fprintf(stderr, "spin_unlock:pthread_mutex_unlock %d\n", ret);
+        abort();
+    }
+}
+
+#define current_tid() (unsigned int) pthread_self()
+
+/* Be careful here, since the C11 terms do no have the same sequential
+ * consistency  for the smp_mb(). Here we use the closely C11 terms,
+ * memory_order_seq_cst.
+ */
+#define smp_mb() atomic_thread_fence(memory_order_seq_cst)
+
+/* Compiler barrier, prevent the compiler to move the memory access from side
+ * to the other side.
+ */
+#define barrier() __asm__ __volatile__("" : : : "memory")
+
+/* Don't use it directly. Use READ_ONCE() or WRITE_ONCE(). */
+#define ACCESS_ONCE(x) (*(volatile __typeof__(x) *) &(x))
```

Comment:
Ensure that you do `#undef ACCESS_ONCE` after the definitions of `READ_ONCE` and `WRITE_ONCE`.

Reply:
Now the `ACCESS_ONCE()` has a `_Atomic` type, and it can be used at atomic operation to solving the `clang` type warning.
But still, without atomic operation, it cannot be used.